api/src/main/java/au/org/aodn/nrmn/restapi/service/SurveyContentsHandler.java

```diff
@@ -0,0 +1,172 @@
+package au.org.aodn.nrmn.restapi.service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import org.apache.poi.xssf.eventusermodel.XSSFSheetXMLHandler.SheetContentsHandler;
+import org.apache.poi.xssf.usermodel.XSSFComment;
+
+import au.org.aodn.nrmn.restapi.dto.payload.ErrorInput;
+import au.org.aodn.nrmn.restapi.model.db.StagedRow;
+import cyclops.control.Validated;
+
+public class SurveyContentsHandler implements SheetContentsHandler {
+
+    private List<String> requiredHeaders;
+
+    private Validated<ErrorInput, List<StagedRow>> result;
+    private StagedRow currentRow;
+    private boolean isHeaderRow = false;
+    private boolean rowHasId = false;
+    private HashMap<String, String> columnHeaders = new HashMap<>();
+    HashMap<Integer, String> measureJson = new HashMap<Integer, String>();
+    List<StagedRow> stagedRows = new ArrayList<StagedRow>();
+
+    SurveyContentsHandler(List<String> requiredHeaders) {
+        this.requiredHeaders = requiredHeaders;
+    }
+
+    public Validated<ErrorInput, List<StagedRow>> getResult() {
+        if (this.result != null)
+            return this.result;
+        else
+            return Validated.invalid(new ErrorInput("DATA sheet not found", "excel"));
+    }
+
+    @Override
+    public void startRow(int rowNum) {
+        rowHasId = false;
+        isHeaderRow = (rowNum == 0);
+        currentRow = StagedRow.builder().pos(rowNum - 1).build();
+    }
+
+    @Override
+    public void endRow(int rowNum) {
+        if (isHeaderRow) {
+            List<String> foundHeaders = new ArrayList<String>(columnHeaders.values());
+            List<String> missingHeaders = new ArrayList<String>(requiredHeaders);
+            missingHeaders.removeAll(foundHeaders);
+            if (missingHeaders.size() > 0)
+                result = Validated
+                        .invalid(new ErrorInput("Missing Headers: " + String.join(", ", missingHeaders), "headers"));
+        } else {
+            if (rowHasId) {
+                if (measureJson.size() > 0)
+                    currentRow.setMeasureJson(new HashMap<Integer,String>(measureJson));
+                this.stagedRows.add(currentRow);
+            }
+        }
+        measureJson.clear();
+    }
+
+    @Override
+    public void endSheet() {
+        if (stagedRows.size() > 0)
+            result = Validated.valid(stagedRows);
+        else
+            result = result != null ? result : Validated.invalid(new ErrorInput("Empty DATA sheet", "sheet"));
+        SheetContentsHandler.super.endSheet();
```

Comment:
Is this right - I would have expected invoking the inherited endSheet on this instance i.e. super.endSheet() 
This is a static reference and so wouldn't have any  state if its required.

Reply:
May be OK but just super.endSheet() is what I'm used to

Reply:
This was boilerplate added by the editor. POI code samples aren't calling this so I'll remove it.